components/servo_tracing/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use quote::{ToTokens, quote};
+use syn::{
+    Attribute, Expr, ItemFn, Meta, MetaList, MetaNameValue, Token,
+    parse::{Parse, Parser},
+    parse_quote, parse2,
+    punctuated::Punctuated,
+    token::Comma,
+};
+
+struct InstrumentConfiguration {
+    name: Option<MetaNameValue>,
+    passthrough_args: Vec<Meta>,
+    field_args: Option<MetaList>,
+    level: Option<Expr>,
+}
+
+impl Parse for InstrumentConfiguration {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let args = Punctuated::<Meta, Token![,]>::parse_terminated(input)?;
+
+        let mut level = None;
+        let mut field_args = None;
+        let mut passthrough_args = vec![];
+        let mut name = None;
+        let span = input.span();
+
+        for arg in args {
+            match arg {
+                Meta::List(meta_list)
+                    if meta_list.path.is_ident("fields") && field_args.is_none() =>
+                {
+                    field_args.replace(meta_list);
```

Comment:
We could check the return value of replace, and do the early return if it was Some (merging the match arm below)